File: app/src/main/java/io/github/lee0701/heonot/KOKR/event/HardKeyEvent.java
package io.github.lee0701.heonot.KOKR.event;

import android.os.Build;
import android.view.KeyEvent;

public class HardKeyEvent extends Event {

	private final HardKeyAction action;

	private final int keyCode;
	private final int repeated;
	private final int metaState;

	public HardKeyEvent(HardKeyAction action, int keyCode, int metaState, int repeated) {
		this.action = action;
		this.keyCode = keyCode;
		this.metaState = metaState;
		this.repeated = repeated;
	}

	public HardKeyAction getAction() {
		return action;
	}

	public int getKeyCode() {
		return keyCode;
	}

	public int getMetaState() {
		return metaState;
	}

	public boolean isShiftPressed() {
		return (metaState & KeyEvent.META_SHIFT_ON) != 0;
	}

	public boolean isAltPressed() {
		return (metaState & KeyEvent.META_ALT_ON) != 0;
	}

	public boolean isCtrlPressed() {
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
			return (metaState & KeyEvent.META_CTRL_ON) != 0;
		} else {
			return false;
		}
	}

	public int getRepeated() {
		return repeated;
	}

	public enum HardKeyAction {
		PRESS, RELEASE, CANCEL
	}

}

File: app/src/main/java/io/github/lee0701/heonot/inputmethod/event/HardKeyEvent.kt
package io.github.lee0701.heonot.inputmethod.event

import android.os.Build
import android.view.KeyEvent


class HardKeyEvent @JvmOverloads constructor(
    private val originalEvent: KeyEvent? = null,
    val action: HardKeyAction,
    val keyCode: Int,
    val metaState: Int,
    val repeated: Int
) : Event() {

    val isShiftPressed: Boolean
        get() = metaState and KeyEvent.META_SHIFT_ON != 0

    val isAltPressed: Boolean
        get() = metaState and KeyEvent.META_ALT_ON != 0

    val isCtrlPressed: Boolean
        get() = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
            metaState and KeyEvent.META_CTRL_ON != 0
        } else {
            false
        }

    enum class HardKeyAction {
        PRESS, RELEASE, CANCEL
    }
}
